### ISSUE TITLE
Integration with Angular locales

### ISSUE THREAD

**Issue by sebastianljunggren** [js-joda/js-joda#384]
I'm currently working on an Angular application that uses js-joda for date parsing, formatting and artithmetic. One problem with this is that Angular already provides prebuilt locale files based on CLDR meaning there is both some duplicated work (setting up locales for js-joda and Angular) and code (two sets of CLDR data is loaded).

Is there any way to load the files Angular provides (example: https://github.com/angular/angular/blob/master/packages/common/locales/sv.ts) into @js-joda/locale with the current library?

If not, would it be feasible to build something like @js-joda/locale-angular that provides a bridge between Angular and  js-joda?

**Comment by phueper** [comment]
Hi @sebastianljunggren 

Thanks for using js-joda :) from what i can see, i don't think that the file you mentioned can be used as is... we are using the clrdjs and cldr-data packages in @js-joda/locale... we do some webpack tricks to replace the cldr-data load mechanism (which is using fs iirc) to just do `require` of the cldr-data json files . If you do have access to these files, it might be possible to do similar tricks to load the data into @js-joda/locale and provide a bridge as you mentioned.

I'm afraid i can't help you much more than that at the moment. 

Feel free to contact us if you have further questions.

Regards, Pattrick

**Comment by sebastianljunggren** [comment]
Thanks for the quick reply! I was kind of expecting something like this but I though I should check anyway.

A follow up question that is slightly related. One of the reasons I'm asking about using Angular locales is that they seem to be much more lightweight. Some of the prebuilt locales (in our case it is @js-joda/locale_fr) are very big. The French package seems to add ~2MB to our bundle size while the more specific ones like de-DE and en-US are much more lightweight.

Is it a lot of work for you to add more locale packages? Would you be open to contributions of more packages (if this would be reasonable for a beginner to do)? Our current wishlist is fr-FR and sv-SE.

**Comment by phueper** [comment]
Hi @sebastianljunggren

First it should be really easy to add more locale packages... Basically just add them here https://github.com/js-joda/js-joda/blob/master/packages/locale/build_package.prebuilt.json... Add the build results in git, make a PR, convince me to merge it and push me to publish them (the latter two might be the hardest part here 😉).. And we'd be very happy to accept contributions. 

As for the other question regarding the locale packages... I'm not very happy with the size either, but we took as the first option cldr since it basically has everything... We are trying to get rid of everything we know we don't need but since we don't know what people really use it is very hard to strip the cldr packages down much more. 
That said I believe we have a pretty good test coverage and I'd be happy to discuss and help in creating packages based on more lightweight locale data and give users a better choice... I honestly don't know how hard it would be to use angular packages but in principal it should be possible and we could create something beside https://github.com/js-joda/js-joda/tree/master/packages/locale/src/format/cldr  using the angular data and try to get that to be used and create more packages... Actually that might give people more choice and would be a good thing i guess... That said, I don't have much time currently to code much myself in js-joda, but I'd be happy to help, discuss,...

Looking forward to contributions... 

Regards, Pattrick

**Comment by sebastianljunggren** [comment]
Ok, we'll see when and if I get the time to investigate contributing more locales. Thanks for your help. You may close this issue since it turned out to be more of a support request.

**Comment by sebastianljunggren** [comment]
I made a quick test trying to add the locales i mentioned (`sv`, `sv-SE` and `fr-FR`).

Then I ran:

```bash
lerna bootstrap
lerna run build-all
```
It seems to build the locales properly but gives me a huge changeset that includes lots of files I did not expect to change. This is the result of `git status` after adding all changes:

```bash
> git status
On branch master
Your branch is up to date with 'origin/master'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   packages/core/dist/js-joda.esm.js
	modified:   packages/core/dist/js-joda.js
	modified:   packages/core/dist/js-joda.js.map
	modified:   packages/core/dist/js-joda.min.js
	modified:   packages/locale/build_package.prebuilt.json
	new file:   packages/locale/dist/prebuilt/deDe/index.js
	new file:   packages/locale/dist/prebuilt/deDe/index.js.map
	new file:   packages/locale/dist/prebuilt/enUs/index.js
	new file:   packages/locale/dist/prebuilt/enUs/index.js.map
	new file:   packages/locale/dist/prebuilt/fr-fr/index.js
	new file:   packages/locale/dist/prebuilt/fr-fr/index.js.map
	new file:   packages/locale/dist/prebuilt/frFr/index.js
	new file:   packages/locale/dist/prebuilt/frFr/index.js.map
	new file:   packages/locale/dist/prebuilt/itIt/index.js
	new file:   packages/locale/dist/prebuilt/itIt/index.js.map
	new file:   packages/locale/dist/prebuilt/sv-se/index.js
	new file:   packages/locale/dist/prebuilt/sv-se/index.js.map
	new file:   packages/locale/dist/prebuilt/sv/index.js
	new file:   packages/locale/dist/prebuilt/sv/index.js.map
	new file:   packages/locale/dist/prebuilt/svSe/index.js
	new file:   packages/locale/dist/prebuilt/svSe/index.js.map
	modified:   packages/locale/packages/de-de/package.json
	modified:   packages/locale/packages/de/package.json
	new file:   packages/locale/packages/deDe/README.md
	new file:   packages/locale/packages/deDe/dist/index.js
	new file:   packages/locale/packages/deDe/dist/index.js.map
	new file:   packages/locale/packages/deDe/package.json
	modified:   packages/locale/packages/en-us/package.json
	modified:   packages/locale/packages/en/package.json
	new file:   packages/locale/packages/enUs/README.md
	new file:   packages/locale/packages/enUs/dist/index.js
	new file:   packages/locale/packages/enUs/dist/index.js.map
	new file:   packages/locale/packages/enUs/package.json
	modified:   packages/locale/packages/es/package.json
	new file:   packages/locale/packages/fr-fr/README.md
	new file:   packages/locale/packages/fr-fr/dist/index.js
	new file:   packages/locale/packages/fr-fr/dist/index.js.map
	new file:   packages/locale/packages/fr-fr/package.json
	modified:   packages/locale/packages/fr/package.json
	new file:   packages/locale/packages/frFr/README.md
	new file:   packages/locale/packages/frFr/dist/index.js
	new file:   packages/locale/packages/frFr/dist/index.js.map
	new file:   packages/locale/packages/frFr/package.json
	modified:   packages/locale/packages/hi/package.json
	modified:   packages/locale/packages/it-it/package.json
	modified:   packages/locale/packages/it/package.json
	new file:   packages/locale/packages/itIt/README.md
	new file:   packages/locale/packages/itIt/dist/index.js
	new file:   packages/locale/packages/itIt/dist/index.js.map
	new file:   packages/locale/packages/itIt/package.json
	new file:   packages/locale/packages/sv-se/README.md
	new file:   packages/locale/packages/sv-se/dist/index.js
	new file:   packages/locale/packages/sv-se/dist/index.js.map
	new file:   packages/locale/packages/sv-se/package.json
	new file:   packages/locale/packages/sv/README.md
	new file:   packages/locale/packages/sv/dist/index.js
	new file:   packages/locale/packages/sv/dist/index.js.map
	new file:   packages/locale/packages/sv/package.json
	new file:   packages/locale/packages/svSe/README.md
	new file:   packages/locale/packages/svSe/dist/index.js
	new file:   packages/locale/packages/svSe/dist/index.js.map
	new file:   packages/locale/packages/svSe/package.json
	modified:   packages/locale/packages/zh/package.json

```
There's a lot of changes that I did not expect here. The `deDe`, `svSe` and so on packages just seem incorrect and there are lots of changes like having spaces removed in the `package.json` files. If I were to make a PR, I'm simply not sure what part I should include. Did I miss something?

**Comment by phueper** [comment]
Hi,

Actually that looks OK the changed files are the build artefacts... I'd recommend you create a PR with just the changed config file, not the build output for now... I can have a look at the changes, they might be due to some dependency updates... 

Regards, Pattrick

**Comment by phueper** [comment]
@sebastianljunggren  ... thanks a lot for the PR... 

i merged it, updated README and packages and published @js-joda/locale* verstion 3.1.1 ... so @js-joda/locale_fr-fr (considerably smaller than the fr* package) as well as @js-joda/locale_sv and @js-joda/locale_sv-se should be published and available now... i'll close this issue now.

But, if you are interested in trying to use the angular locale packages, the offer stands, i'd be happy to discuss and support this, although it will be considerably more work than this change :) 

(also, i'd be happy to hear wether this helped you and helped your project )

**Comment by sebastianljunggren** [comment]
Thank you for all your help. I have updated our configs and the Swedish locale seems to work and for French we have reduced our bundle size by ~2MB and no longer get warnings like `WARNING in budgets: Exceeded maximum budget for initial-es2015. Budget 2 MB was not met by 1.27 MB with a total of 3.27 MB.` when building. This is really important to our project and has really simplified things for my team.

Since you seem interested, I'll share some info about our experience of using js-joda and an idea for how locales could be supported by the library.

# Our experience with js-joda

> (also, i'd be happy to hear wether this helped you and helped your project)

Since you are asking, I'll give a description (mostly praise), of what our experience of working with js-joda has been like.

My team is building an Angular application, of what I would describe as a medium size, in a domain where users are shown dates in different timezones in a single view. Therefore, getting the conversions and formats correct is really important. The distinction  between `LocalDateTime`, `Instant` and `ZonedDateTime`is also critical for the same reason in our code base.

Previously, we have been using Moment.js for our date needs but for various reasons we have ended up in a situation where it was really hard to know what type of data (local, zoned, instant) a specific moment represented and how it should be converted to a different format. After starting to use js-joda in combination with TypeScript reasoning about dates has become a lot easier for us. The fact that many of the developers in the team have a Java background and that we make heavy use of `java.time` has also made the choice of js-joda easy for us.

We are pretty much making use of the entire js-joda ecosystem, including most of the base time and date types, durations, formatting, building for multiple locales and making use of timezone data.

Historically, we have had a lot of time zone and date conversion related bugs, and while we don't have proof yet I believe that this will go down a lot when we go live with our new front end built with js-joda and Angular.

However, there has been a few gaps we have encountered in our work with js-joda:

- No formatting for durations like in moment-duration-format. Solved by creating a basic formatting function that meets our needs (at some point this may be released as a separate lib on npm).
- Including a timzone database in our front end build is not great for the bundle size and that db may be out of sync with the back end. Currently solved by using one of the smaller databases that you provide (thanks!) but we plan to refactor our backend to only return `LocalDateTime` and `OffsetDateTime`so we can skip including the DB at all in the future.
- Not being able to use multiple locales at the same time means that we currently only are able to run unit tests on our formatting service with our default (English) locale.
- The locale issues that have been discussed in this issue

# Suggestion for how to support multiple locales in js-joda

Regarding using the Angular locales, I've been doing some thinking on how js-joda could/should handle locale data. This is just an idea, so please view this as feedback and not criticism.

To me there seems to be two maintenance related problems with the current way of handling locale data:

- The repo grows big since a lot of data is checked into Git. It already takes a while to clone the repo. This problem will become worse for each "big locale" you add.
- Having one package per locale does not really scale if you were to support a lot of locales like Angular and many other projects do.

I've had a look at what the Angular project does and I think that approach could work for js-joda as well. They provide locale files with the things that 90%  (not an actual stat but a rough estimation by me) of their users need. For js-joda I believe that could be something like:

- Names of days
- Names of months
- AM/PM
- BC/AD
- The default date formats if they are easy to translate to js-joda format strings

For users that want more, like names of time zones, the option of generating CLDR data would still be available.

Instead of supporting a few locales with a lot of data, js-joda could instead focus on supporting many locales with only the base data that most users need. COmparing to Angular again, they seem to generate their data for all CLDR locales (see https://github.com/angular/angular/blob/master/tools/gulp-tasks/cldr/extract.js#L78 for how they generate the locales and https://github.com/angular/angular/tree/master/packages/common/locales for the locales they have generated). This data could be inluded in the `@js-joda/locale`package and the existing locale specific packages could possibly be deprecated. Perhaps you could load the locale data in something like this code sample:

```js
import { loadLocaleData } from '@js-joda/locale';
import { data as de } from '@js-joda/locale/de';
import { data as sv } from '@js-joda/locale/sv';

loadLocaleData(de);
loadLocaleData(sv);
```
All of this is of course suggested from the perspective of someone who builds for the browser where bundle size is critical. :)

To summarize, this would kind of be the inverse of the current situation, the responsiblity of the lib would be to provide minimal locale packages instead of full locale packages. Users with more specific needs would need a custom build instead of those with who have more basic needs or want minimal bundles.

Once more, thank you for all your help!

**Comment by pithu** [comment]
Hello @sebastianljunggren 

Yes would be great, if any combination of locales could be imported, we already discussed that, but unfortunately didn't find time to implement it. But of course we are open for PR on that.

I would even simplify it to something like the following. In the end it goes into a singleton anyway:

```
import { Locale } from '@js-joda/locale';
import  '@js-joda/locale/de';
import  '@js-joda/locale/sv';
```
